publisher-sdk/src/main/java/com/criteo/publisher/util/jsonadapter/ArrayListJsonAdapter.kt

```diff
@@ -0,0 +1,108 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher.util.jsonadapter
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import java.io.IOException
+import java.lang.reflect.Type
+import java.util.Arrays
+
+internal abstract class ArrayListJsonAdapter<C : MutableCollection<T>?, T> private constructor(
+    private val elementAdapter: JsonAdapter<T>
+) : JsonAdapter<C>() {
+  abstract fun newCollection(): C
+
+  @Throws(IOException::class)
+  override fun fromJson(reader: JsonReader): C {
+    val result = newCollection()
+    reader.beginArray()
+    while (reader.hasNext()) {
+      result?.add(elementAdapter.fromJson(reader)!!)
+    }
+    reader.endArray()
+    return result
+  }
+
+  @Throws(IOException::class)
+  override fun toJson(writer: JsonWriter, value: C?) {
+    writer.beginArray()
+    for (element in value!!) {
+      elementAdapter.toJson(writer, element)
+    }
+    writer.endArray()
+  }
+
+  override fun toString(): String {
+    return "$elementAdapter.collection()"
+  }
+
+  companion object {
+    @JvmStatic
+    val factory = Factory { type, annotations, moshi ->
+      val rawType = Types.getRawType(type)
+      if (annotations.isNotEmpty()) return@Factory null
+      // not really the best way to compare but class is private
+      if (rawType.name == "java.util.Arrays${'$'}ArrayList") {
+        return@Factory newArraysArrayListAdapter<Any>(
+            type,
+            moshi
+        ).nullSafe()
+      }
+      if (rawType == java.util.ArrayList::class.java) {
+        return@Factory newArrayListAdapter<Any>(
+            type,
+            moshi
+        ).nullSafe()
+      }
+      null
+    }
+
+    private fun <T> newArraysArrayListAdapter(
+        type: Type,
+        moshi: Moshi
+    ): JsonAdapter<MutableCollection<T>> {
+      val elementType = Types.collectionElementType(type, Collection::class.java)
+      val elementAdapter: JsonAdapter<T> = moshi.adapter(elementType)
+
+      return object :
+          ArrayListJsonAdapter<MutableCollection<T>, T>(elementAdapter) {
+        override fun newCollection(): MutableCollection<T> {
+          return Arrays.asList()
+        }
+      }
+    }
+
+    private fun <T> newArrayListAdapter(
+        type: Type,
+        moshi: Moshi
+    ): JsonAdapter<MutableCollection<T>> {
+      val elementType = Types.collectionElementType(type, Collection::class.java)
+      val elementAdapter: JsonAdapter<T> = moshi.adapter(elementType)
+
+      return object :
+          ArrayListJsonAdapter<MutableCollection<T>, T>(elementAdapter) {
+        override fun newCollection(): MutableCollection<T> {
+          return arrayListOf()
```

Comment:
Why not just using the Java constructor directly ?

Reply:
Yep, will change that